e2e/process_test.go

```diff
@@ -0,0 +1,469 @@
+		initialToken: "new",
+	}
+	defaultConfigTLS = etcdProcessClusterConfig{
+		clusterSize:  3,
+		proxySize:    0,
+		isClientTLS:  true,
+		isPeerTLS:    true,
+		initialToken: "new",
+	}
+	defaultConfigClientTLS = etcdProcessClusterConfig{
+		clusterSize:  3,
+		proxySize:    0,
+		isClientTLS:  true,
+		isPeerTLS:    false,
+		initialToken: "new",
+	}
+	defaultConfigPeerTLS = etcdProcessClusterConfig{
+		clusterSize:  3,
+		proxySize:    0,
+		isClientTLS:  false,
+		isPeerTLS:    true,
+		initialToken: "new",
+	}
+	defaultConfigWithProxy = etcdProcessClusterConfig{
+		clusterSize:  3,
+		proxySize:    1,
+		isClientTLS:  false,
+		isPeerTLS:    false,
+		initialToken: "new",
+	}
+	// TODO: this does not work now
+	defaultConfigWithProxyTLS = etcdProcessClusterConfig{
+		clusterSize:  3,
+		proxySize:    1,
+		isClientTLS:  true,
+		isPeerTLS:    true,
+		initialToken: "new",
+	}
+)
+
+type etcdProcessCluster struct {
+	cfg   *etcdProcessClusterConfig
+	procs []*etcdProcess
+}
+
+type etcdProcess struct {
+	cfg   *etcdProcessConfig
+	proc  *gexpect.ExpectSubprocess
+	donec chan struct{} // closed when Interact() terminates
+}
+
+type etcdProcessConfig struct {
+	args        []string
+	name        string
+	dataDirPath string
+	purl        url.URL
+	acurl       url.URL
+	isProxy     bool
+}
+
+type etcdProcessClusterConfig struct {
+	clusterSize  int
+	proxySize    int
+	isClientTLS  bool
+	isPeerTLS    bool
+	initialToken string
+}
+
+// newEtcdProcessCluster launches a new cluster from etcd processes, returning
+// a new etcdProcessCluster once all nodes are ready to accept client requests.
+func newEtcdProcessCluster(cfg *etcdProcessClusterConfig) (*etcdProcessCluster, error) {
+	etcdCfgs := cfg.etcdProcessConfigs()
+	epc := &etcdProcessCluster{
+		cfg:   cfg,
+		procs: make([]*etcdProcess, cfg.clusterSize+cfg.proxySize),
+	}
+
+	// launch etcd processes
+	for i := range etcdCfgs {
+		proc, err := newEtcdProcess(etcdCfgs[i])
+		if err != nil {
+			epc.Close()
+			return nil, err
+		}
+		epc.procs[i] = proc
+	}
+
+	// wait for cluster to start
+	readyC := make(chan error, cfg.clusterSize+cfg.proxySize)
+	readyStr := "etcdserver: set the initial cluster version to"
+	for i := range etcdCfgs {
+		go func(etcdp *etcdProcess) {
+			rs := readyStr
+			if etcdp.cfg.isProxy {
+				rs = "proxy: listening for client requests on"
+			}
+			ok, err := etcdp.proc.ExpectRegex(rs)
+			if err != nil {
+				readyC <- err
+			} else if !ok {
+				readyC <- fmt.Errorf("couldn't get expected output: '%s'", rs)
+			} else {
+				readyC <- nil
+			}
+			etcdp.proc.ReadLine()
+			etcdp.proc.Interact() // this blocks(leaks) if another goroutine is reading
+			etcdp.proc.ReadLine() // wait for leaky goroutine to accept an EOF
+			close(etcdp.donec)
+		}(epc.procs[i])
+	}
+	for range etcdCfgs {
+		if err := <-readyC; err != nil {
+			epc.Close()
+			return nil, err
+		}
+	}
+	if epc.cfg.proxySize > 0 {
+		for i := 0; i < 5; i++ {
+			ok, _ := isProxyReady(epc)
```

Comment:
Yes, it prints that out before serving client requests: https://github.com/coreos/etcd/blob/master/etcdmain/etcd.go#L462-L466.